veps/sig-storage/incremental-backup.md

```diff
@@ -0,0 +1,374 @@
+
+## Non Goals
+
+Have an API to restore from backup. In this VEP we will present a way to do it but its not something that we will have an API for.
+
+## Definition of Users
+
+* Backup vendors
+* Cluster Admins
+* VM owners
+
+## User Stories
+
+* As a KubeVirt user, I would like to back up my VM in the most efficient way possible, both in terms of time and storage usage.
+* As a Kubevirt admin I would like to take a complete backup of my cluster. Then, I would like to take backup of the changes since the previous backup.
+* As a Kubevirt admin I would like to store the backups off-site and lower the amount of data I am copying each time.
+* As a Kubevirt admin I would like to restore to a specific time by applying both base backup and smaller time-specific increment.
+
+## Repos
+
+[KubeVirt](https://github.com/kubevirt/kubevirt)
+
+## Design
+
+### Enable/Disable QEMU backup
+
+QEMU backup with CBT is only supported with QCOW2 images, currently, Kubevirt supports only raw images.
+To enable QEMU backup for a disk, we need to create a QCOW2 overlay that will store the **image metadata** (Not data!) and enable the use of QCOW2 features.
+This overlays will be stored on a VM state PVC, similar to how TPM and EFI are handled. If the VM uses either of these features, the overlays will be created on the same PVC. Figure 1 describes how this will look like:
+![figure 1](qcow2overlay.png)
+
+A new field, `changedBlockTracking`, will be added to the VirtualMachine CRD. This field will control the addition or removal of the QCOW2 overlay for each of the supported VM volumes (i.e DataVolumes and PersistentVolumeClaims).
+This field will be set to `true` either:
+* Automatically, when the first backup is initiated.
+* Automatically, upon creation of `VirtualMachineBackupChain` CR (will be explained in next section)
+* Manually, by the user — helps avoid a restart right before the first backup and reduces the time required to initiate it.
+
+Currently, live addition of the overlay is not supported and the Libvirt team is working on enabling it. Once it's available, we will work on incorporating it into KubeVirt as well. Until then, a VM restart will be required for any changes to the `changedBlockTracking` field to take effect and be reflected in the VMI.
+
+When `changedBlockTracking` set to `true` in the VMI spec:
+1. VM state PVC will be created (if one doesn’t already exist).
+2. QCOW2 image overlay will be created on top of the relevant disks.
+3. A `data-store` tag will be added to each of those disks in the domain XML.
+As a result, the domain will be configured with QCOW2 overlays.
+Once that is done the VMI volume status will be updated with changedBlockTracking: `true` for the relevant volumes.
+*NOTE: In the case of a hot-plugged disk, the VM will need to be restarted for it to be included in the backup and have CBT. Once live addition of the overlay is supported, restarts will no longer be required.
```

Comment:
What happens when backup is created with hotplug volume and no overlay? I assume the backup will proceed but there will be some indication in the backup status?

Reply:
Hmm, well in the volume status of the vmi you can see which disks has the overlay.
Do we want to have `includedVolumes` list similar to in vmsnapshot that will indicate which volumes were included? 
Also in the checkpoints list info if it will not be encoded you can see that xml of the checkpoint which states which disks were included in the backup, but not sure if it will not make the backupChain status too big

Reply:
As discussed offline, we dont have the problem of a restart with hotplug since we can create the overlay dynamically when hotplugging the volume, so no need for this field or indication